---
title: "Differential expression analysis of ROSMAP RNASeq data based on ADNI 1.5T 3Yr results from Elias"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

# source('/mnt/Github/knit2synapse-1/R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_ROSMAP_RNASeq.Rmd", owner = 'syn4239880', name = "Differential Expression Analysis ROSMAP RNASeq",overwrite=F)

## Load required libraries
library(synapseClient)
library(ggplot2)
library(reshape2)
library(data.table)
library(dplyr)
library(stringr)
library(limma)
library(knitr)
library(Biobase)
library(GEOquery)
library(RColorBrewer)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

## Needs the dev branch
library(rGithubClient)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn3721177';
activityName = 'Differential expression analysis';
activityDescription = 'Differential expression analysis based on image features';

thisFileName <- 'DiffExpAnal_ROSMAP_RNASeq.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/Takeda_contract", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('Rmd/',thisFileName))
```
### Image Analysis
1. Image features used:
  * FreeSurfer.thickness..median  
  * mean.curvature..median         
  * Volume
  * travel.depth..median  
  * FreeSurfer.convexity..sulc...median
```{r load.data, cache=TRUE}
adni.mem.1_5T.3Y <- read.csv('./adni.mem_ranked_1.5T_3Yr_takeda_4_6_15.csv')
label.name.conversion <- read.csv('./label_names_conversion.csv')

adni.mem.1_5T.3Y <- dplyr::mutate(adni.mem.1_5T.3Y, 
                                  sig.feature.all = a.ks.pvals <= 0.05 & a.cor.pvals <= 0.05,
                                  sig.feature.aks = a.ks.pvals <= 0.05,
                                  feature = sapply(str_split(adni.mem.1_5T.3Y$var.nms,'\\.'),function(x){paste(x[-(length(x))],collapse='.')}),
                                  region = sapply(str_split(adni.mem.1_5T.3Y$var.nms,'\\.'),function(x){paste(x[(length(x))],collapse='.')}))
adni.mem.1_5T.3Y <- merge(adni.mem.1_5T.3Y,label.name.conversion, by.x = 'region', by.y = 'Label')
sig.table.all <- xtabs(sig.feature.all ~ feature + region,data = adni.mem.1_5T.3Y)
colnames(sig.table.all) <- label.name.conversion[match(label.name.conversion$Label,colnames(sig.table.all)),1]

unique.features <- unique(sapply(str_split(rownames(sig.table.all),'\\.'),function(x){x <- x[x!=""];l <- length(x);ifelse(l>1,paste(x[-(l)],collapse='.'),x)}))

region.ranking <- colSums(sig.table.all[rownames(sig.table.all) %in% c('FreeSurfer.thickness..median','mean.curvature..median','Volume','travel.depth..median','FreeSurfer.convexity..sulc...median'),])
region.ranking <- sort(region.ranking,decreasing = T)
region.ranking <- region.ranking[region.ranking!=0]

unique.regions.correlated <- unique(sapply(str_split(names(region.ranking),'-'),function(x){x[[length(x)]]}))

sig.table.aks <- xtabs(sig.feature.aks ~ feature + Label.name, data = adni.mem.1_5T.3Y)
region.ranking <- colSums(sig.table.aks[rownames(sig.table.aks) %in% c('FreeSurfer.thickness..median','mean.curvature..median','Volume','travel.depth..median','FreeSurfer.convexity..sulc...median'),])
region.ranking <- sort(region.ranking,decreasing = T)
region.ranking <- region.ranking[region.ranking!=0]
unique.regions <- unique(sapply(str_split(names(region.ranking),'-'),function(x){x[[length(x)]]}))

```
2. Significantly changed regions are: `r unique.regions`
3. Significantly changed regions that are also correlated with ADNI.MEM are: `r unique.regions.correlated`
4. Dorsolateral prefrontal cortex of 724 subjects from the ROS and MAP cohorts were obtained from AMP-AD portal

### Data download
Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_ID <- 'syn3505720';
ALL_USED_IDs <- COUNT_ID
COUNT_OBJ <- synGet(COUNT_ID)
COUNT <- read.table(COUNT_OBJ@filePath,header=T,sep='\t',check.names = F)

# COUNT_ID1 <- 'syn3505720'; #'syn3505732'
# ALL_USED_IDs <- COUNT_ID1
# COUNT_OBJ1 <- synGet(COUNT_ID1)
# COUNT1 <- read.table(COUNT_OBJ1@filePath,header=T,sep='\t',check.names = F)
# 
# # Download expression data
# COUNT_ID2 <- 'syn3505724'
# ALL_USED_IDs[length(ALL_USED_IDs)+1] <- COUNT_ID2
# COUNT_OBJ2 <- synGet(COUNT_ID2)
# COUNT2 <- read.table(COUNT_OBJ2@filePath,header=T,sep='\t',check.names = F)

# Get clinical metadata
METADATA.CLINICAL_ID <- 'syn3191087'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA.CLINICAL_ID
METADATA.CLINICAL_OBJ <- synGet(METADATA.CLINICAL_ID)
METADATA.CLINICAL <- read.table(METADATA.CLINICAL_OBJ@filePath,sep=',',header=T)

# Get technical covariates
METADATA.TECH <- read.csv('./ROSMAP_RNA_Seq_RIN_Sample_READS.csv')
METADATA.TECH <- METADATA.TECH[!duplicated(METADATA.TECH$Projid),]

# Match technical and clinical covariates
ind <- intersect(METADATA.TECH$Projid,METADATA.CLINICAL$projid)
METADATA.TECH <- METADATA.TECH[METADATA.TECH$Projid %in% ind,]
METADATA.CLINICAL <- METADATA.CLINICAL[METADATA.CLINICAL$projid %in% ind,]
METADATA <- merge(METADATA.TECH,METADATA.CLINICAL,by.x='Projid',by.y='projid',all=T)
```

### Data preprocessing
```{r preprocess.data,cache=TRUE, echo=TRUE}
# Remove samples with no cogdx
METADATA <- METADATA %>%
  filter(!is.na(cogdx)) %>%
  filter(!is.na(RINcontinuous)) %>%
  filter(!is.na(pmi))
```

```{r preprocess.data1,cache=TRUE}
EXPR <- COUNT; #merge(COUNT1,COUNT2,by = 'gene_id',all=T)
rownames(EXPR) <- EXPR$gene_id
EXPR <- dplyr::select(EXPR,-gene_id,-tracking_id)

# Match covariates to expression data
METADATA <- dplyr::mutate(METADATA, user.id = paste(Sampleid,Batch,sep='_'))
ind <- match(METADATA$user.id,colnames(EXPR))
EXPR <- EXPR[,ind]
```

### Covariate clustering
Determine relationship between covariates. 
```{r covariates.clustering, cache=TRUE}
#"braaksc","ceradsc",,"cts_mmse30_first_ad_dx","cts_mmse30_lv"
FactorCovariates <- c("Batch","msex","apoe_genotype", "race", "spanish", "cogdx")
ContCovariates <- c("RINcontinuous","age_first_ad_dx2","age_death2","age_at_visit_max2","pmi","educ")
  
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES <- data.frame(lapply(COVARIATES,function(x){x <- sapply(x,function(y){str_replace_all(as.character(y),'\\+','')})}))
rownames(COVARIATES) <- METADATA$Projid
colnames(EXPR) <- METADATA$Projid

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Covariate correlation
```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

### Gene Filtering
Retain genes with atleast 50% of the samples have >= 0 FPKM counts
```{r gene.filtering, cache=TRUE}
# Remove genes
EXPR <- EXPR[(rowSums(EXPR>0)/dim(EXPR)[2] >= 0.5),]

# To avoid taking log of zero add 1e-4 to the FPKM counts
EXPR <- log2(EXPR+1e-3)
```
`r dim(EXPR)[1]` number of genes and `r dim(EXPR)[2]` number of samples are considered for the analysis

### Significant Covariates
Correlation between pca of normalised mRNA expression and covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
DM = getDesignMatrix(COVARIATES, Intercept = F)
DESIGN = getDesignMat2Fact(DM$design,FactorCovariates)
rownames(DESIGN) = rownames(COVARIATES)
preAdjustedSigCovars = runPCAandPlotCorrelations(EXPR, DESIGN,'non-normalized', isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),preAdjustedSigCovars$significantCovars)
```
Significant covariates to adjust at FDR 0.1 are `r adjustCovars`
```{r preAdjustedSigCovars.NULL, fig.width=15, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r preAdjustedSigCovars.NULL.ALL, fig.width=15, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```
### Normalisation (custom design)
Fitting a linear model with following variables cogdx, Batch, RINcontinuous, pmi
```{r iterative.normalisation, cache=TRUE}
postAdjustCovars <- c('cogdx','Batch',"RINcontinuous","pmi")

tmp <- paste('Using following covariates in the final model',paste(postAdjustCovars,collapse=','),'as fixed effects',sep=' ')

# Post adjusted design matrix
DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
# Re-calculate linear fit using new weights and new design
FIT = lmFit(EXPR, design=DM1$design)

# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,EXPR)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
rownames(DESIGN) <- rownames(COVARIATES)
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION,
                                              DESIGN,
                                              'all adjusted design',
                                              isKeyPlot=TRUE)
```
`r tmp`

```{r residualSigCovars.manual, fig.width=15, fig.height=8}
residualSigCovars[["PC_res"]][[1]]$plotData
```

```{r residualSigCovars.manual.ALL, fig.width=15, fig.height=8}
residualSigCovars[["PC_res"]][[2]]$plotData
```
 
### Differential Expression
```{r contrast.fit, cache=TRUE, fig.height=10, fig.width=15}
# Make contrast to check differential expression
CONT.NAMES <- colnames(DM1$design)[grep('cogdx',colnames(DM1$design))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DEXP <- list()
DEXP$logFC <- data.frame(row.names = rownames(EXPR))
DEXP$adj.P.Val <- data.frame(row.names = rownames(EXPR))
  
for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR)[1])    
  DEXP$logFC[,i] <- tmp[rownames(DEXP$logFC),'logFC']
  DEXP$adj.P.Val[,i] <- tmp[rownames(DEXP$adj.P.Val),'adj.P.Val']
}
  
# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and abs(logFC) >= 0
DEXP$SIG.EXP <- DEXP$adj.P.Val<=0.05 & abs(DEXP$logFC) >= 0   
DEXP$NUM.SIG.EXP <- colSums(DEXP$SIG.EXP)
```
Interpretation of cogdx scores

1.  NCI, No cognitive impairment (No impaired domains)
2.  MCI, Mild cognitive impairment (One impaired domain) and NO other cause of CI
3.  MCI, Mild cognitive impairment (One impaired domain) AND another cause of CI
4.  AD, Alzheimer's disease and NO other cause of CI (NINCDS PROB AD)
5.  AD, Alzheimer's disease AND another cause of CI (NINCDS POSS AD)
6.  Other dementia. Other primary cause of dementia

```{r print.fc.pval}
# Print results
tmp <- as.data.frame(DEXP$NUM.SIG.EXP)
colnames(tmp) <- 'Diff exp probes at FDR <= 0.05'
kable(tmp)

# Get fold change and pvalues
FC <- rownameToFirstColumn(DEXP$logFC,"EnsembleID")
PVAL <- rownameToFirstColumn(DEXP$adj.P.Val,"EnsembleID")
```

### Store files in synapse
 
```{r synapse.store, include=FALSE, eval=TRUE}
# Code
CODE <- File(paste0('./',thisFileName), name = "Differential Expression Analysis ROSMAP RNASeq", parentId = parentId)
CODE <- synStore(CODE, used = ALL_USED_IDs, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
      
# Fold change
FC <- FC[, unique(colnames(FC))]
write.table(FC, file='./FoldChange_ROSMAP_RNASeq.tsv', sep='\t', row.names=F, quote=F)
FC_OBJ <- File('./FoldChange_ROSMAP_RNASeq.tsv',name = 'Differential Expression Analysis ROSMAP RNASeq logFC',parentId = parentId)
FC_OBJ <- synStore(FC_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id, activityDescription = activityDescription)
    
# Pvalue
PVAL <- PVAL[, unique(colnames(PVAL))]
write.table(PVAL, file='./AdjustedPval_ROSMAP_RNASeq.tsv', sep='\t', row.names=F, quote=F)
PVAL_OBJ <- File('./AdjustedPval_ROSMAP_RNASeq.tsv',name = 'Differential Expression Analysis ROSMAP RNASeq adjPval',parentId = parentId)
PVAL_OBJ <- synStore(PVAL_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id, activityDescription = activityDescription)
```
|  *Results*           |  *SynapseID*                    |
|  -------             |   ---------                     |
|  Log fold change     |  `r FC_OBJ$properties$id`       |
|  Adjusted pvalues    |  `r PVAL_OBJ$properties$id`     |